helpers/chatHelper.py

```diff
@@ -266,462 +264,629 @@ async def partChannel(
+                "user_id": user_token["user_id"],
+                "recipient": send_to,
+            },
+        )
+        return ...
+
+    if await osuToken.isSilenced(recipient_token["token_id"]):
+        await osuToken.enqueue(
+            user_token["token_id"],
+            serverPackets.targetSilenced(recipient_token["username"]),
+        )
+
+        logger.warning(
+            "User tried to send a message to a silenced user",
+            extra={
+                "token_id": user_token["token_id"],
+                "username": user_token["username"],
+                "user_id": user_token["user_id"],
+                "recipient_username": recipient_token["username"],
+                "recipient_user_id": recipient_token["user_id"],
+            },
+        )
+
+        return ...
+
+    if osuToken.is_restricted(recipient_token["privileges"]):
+        logger.warning(
+            "User attempted to send a message to a restricted user",
+            extra={
+                "token_id": user_token["token_id"],
+                "username": user_token["username"],
+                "user_id": user_token["user_id"],
+                "recipient_username": recipient_token["username"],
+                "recipient_user_id": recipient_token["user_id"],
+            },
+        )
+        return ...
+
+    # TODO: check if user has us blacklisted
+    # fmt: off
+    if (
+        recipient_token["block_non_friends_dm"]
+        and user_token["user_id"]
+        not in await userUtils.getFriendList(recipient_token["user_id"])
+    ):
+    # fmt: on
+        await osuToken.enqueue(
+            user_token["token_id"],
+            serverPackets.targetBlockingDMs(recipient_token["username"]),
+        )
+
         logger.warning(
             "User tried to send a message to a user that is blocking non-friends dms",
             extra={
-                "username": userToken["username"],
-                "user_id": userToken["user_id"],
-                "recipient": to,
+                "token_id": user_token["token_id"],
+                "username": user_token["username"],
+                "user_id": user_token["user_id"],
+                "recipient_username": recipient_token["username"],
+                "recipient_user_id": recipient_token["user_id"],
+            },
+        )
+
+        return ...
+
+    if await osuToken.awayCheck(
+        recipient_token["token_id"],
+        user_token["user_id"],
+    ):
+        await enqueue_private_message(
+            user_token=recipient_token,
+            recipient_token=user_token,
+            message=f"\x01ACTION is away: {recipient_token['away_message'] or ''}\x01",
+        )
+
+    await enqueue_private_message(
+        user_token=user_token,
+        recipient_token=recipient_token,
+        message=message,
+    )
+
+
+async def handle_bot_message(
+    user_token: osuToken.Token,
+    send_to: str,
+    message: str,
+) -> Optional[int]:
```

Comment:
i think it's best if we decouple the chat handling from the irc protocol/keep it generic -- we can do that by e.g. returning an error enum here and later mapping those errors to irc status codes

Reply:
holy shit:

![image](https://i.cmyui.xyz/OCuY_vmLKCtbAILR.png)

Reply:
nailed it down to
```py
class ChatMessageError(str, Enum):
    UNKNOWN_CHANNEL = "UNKNOWN_CHANNEL"
    NO_CHANNEL_MEMBERSHIP = "NO_CHANNEL_MEMBERSHIP"
    USER_NOT_FOUND = "USER_NOT_FOUND"
    INSUFFICIENT_PRIVILEGES = "INSUFFICIENT_PRIVILEGES"
    SENDER_SILENCED = "SENDER_SILENCED"
    RECIPIENT_SILENCED = "RECIPIENT_SILENCED"
    SENDER_RESTRICTED = "SENDER_RESTRICTED"
    RECIPIENT_RESTRICTED = "RECIPIENT_RESTRICTED"
    BLOCKED_BY_RECIPIENT = "BLOCKED_BY_RECIPIENT"
    INVALID_MESSAGE_CONTENT = "INVALID_MESSAGE_CONTENT"
```